pkg/bpf/collection.go

```diff
@@ -0,0 +1,251 @@
+// classifyProgramTypes sets the type of ProgramSpecs which the library cannot
+// automatically classify due to them being in unrecognized ELF sections. Only
+// programs of type UnspecifiedProgram are modified.
+//
+// Cilium uses the iproute2 X/Y section name convention for assigning programs
+// to prog array slots, which is also not supported.
+func classifyProgramTypes(spec *ebpf.CollectionSpec) {
+	// Assign a program type based on the first recognized function name.
+	var t ebpf.ProgramType
+	for name := range spec.Programs {
+		switch name {
+		// bpf_xdp.c
+		case "bpf_xdp_entry":
+			t = ebpf.XDP
+		case
+			// bpf_lxc.c
+			"handle_xgress", "handle_to_container",
+			// bpf_host.c
+			"from_netdev", "from_host", "to_netdev", "to_host",
+			// bpf_network.c
+			"from_network",
+			// bpf_overlay.c
+			"to_overlay", "from_overlay":
+			t = ebpf.SchedCLS
+		default:
+			continue
+		}
+
+		break
+	}
+
+	for _, p := range spec.Programs {
+		if p.Type == ebpf.UnspecifiedProgram {
+			p.Type = t
+		}
+	}
+}
+
+// inlineGlobalData replaces all map loads from a global data section with
+// immediate dword loads, effectively performing those map lookups in the
+// loader. This is done for compatibility with kernels that don't support
+// global data maps yet.
+//
+// Currently, all map reads are expected to be 32 bits wide until BTF MapKV
+// can be fully accessed by the caller, which would allow for querying value
+// widths.
+//
+// This works in conjunction with the __fetch macros in the datapath, which
+// emit direct array accesses instead of memory loads with an offset from the
+// map's pointer.
+func inlineGlobalData(spec *ebpf.CollectionSpec) error {
+	data := globalData(spec)
+	if data == nil {
+		// No static data, nothing to replace.
+		return nil
+	}
+
+	// Don't attempt to create an empty map .bss in the kernel.
+	delete(spec.Maps, ".bss")
+
+	for _, prog := range spec.Programs {
+		for i, ins := range prog.Instructions {
+			if !ins.IsLoadFromMap() || ins.Src != asm.PseudoMapValue {
+				continue
+			}
+
+			// The compiler inserts relocations for .bss for zero values.
+			if ins.Reference == ".bss" {
+				prog.Instructions[i] = asm.LoadImm(ins.Dst, 0, asm.DWord)
+				continue
+			}
+
+			if ins.Reference != globalDataMap {
+				return fmt.Errorf("global constants must be in %s, but found reference to %s", globalDataMap, ins.Reference)
+			}
+
+			// Get the offset of the read within the target map,
+			// stored in the 32 most-significant bits of Constant.
+			// Equivalent to Instruction.mapOffset().
+			off := uint32(uint64(ins.Constant) >> 32)
+
+			if off%4 != 0 {
+				return fmt.Errorf("global const access at offset %d not 32-bit aligned", off)
+			}
+
+			imm := spec.ByteOrder.Uint32(data[off : off+4])
+
+			// Replace the map load with an immediate load. Must be a dword load
+			// to match the instruction width of a map load.
+			prog.Instructions[i] = asm.LoadImm(ins.Dst, int64(imm), asm.DWord)
+		}
+	}
+
+	return nil
```

Comment:
If there are unused vars, the compiler should warn/error. If there are globals marked `unused`, that's usually for a reason, so wouldn't this kind of warning just add noise?

I've found with this kind of code it's better to either error hard, or automatically do the right thing instead of throwing a warning. There's so much output these tend to get lost.

Reply:
:+1: Fair enough yeah. Agree around automatically doing the right thing. My initial thought was that I couldn't think of any reasons for `If there are globals marked unused, that's usually for a reason`, but now I think more it seems like it could be a convenient way to just stuff some data in an ELF if you need it. So no need for any follow up here.